codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java

```diff
@@ -307,22 +310,45 @@ private void encodeFullHttpMessage(ChannelHandlerContext ctx, Object o, List<Obj
 
             final H m = (H) o;
 
-            final ByteBuf buf = ctx.alloc().buffer((int) headersEncodedSizeAccumulator);
-
-            encodeInitialLine(buf, m);
-
             final int state = isContentAlwaysEmpty(m) ? ST_CONTENT_ALWAYS_EMPTY :
                     HttpUtil.isTransferEncodingChunked(m) ? ST_CONTENT_CHUNK : ST_CONTENT_NON_CHUNK;
 
+            ByteBuf content = msg.content();
+
+            final boolean copyContent = content.readableBytes() > 0 &&
+                                        state == ST_CONTENT_NON_CHUNK &&
+                                        // try embed the content if less or equals than
+                                        // the biggest of ~12.5% of the header estimated size and COPY_DATA_THRESHOLD:
+                                        // it limits a wrong estimation to waste too much memory
+                                        content.readableBytes() <=
+                                        Math.max(COPY_CONTENT_THRESHOLD, ((int) headersEncodedSizeAccumulator) / 8);
+
+            final int headersAndContentSize = (int) headersEncodedSizeAccumulator +
+                                                  (copyContent? content.readableBytes() : 0);
+            final ByteBuf buf = ctx.alloc().buffer(headersAndContentSize);
+
+            encodeInitialLine(buf, m);
+
             sanitizeHeadersBeforeEncode(m, state == ST_CONTENT_ALWAYS_EMPTY);
 
             encodeHeaders(m.headers(), buf);
             ByteBufUtil.writeShortBE(buf, CRLF_SHORT);
 
+            // don't consider the copyContent case here: the statistics is just related the headers
             headersEncodedSizeAccumulator = HEADERS_WEIGHT_NEW * padSizeForAccumulation(buf.readableBytes()) +
                     HEADERS_WEIGHT_HISTORICAL * headersEncodedSizeAccumulator;
 
-            encodeByteBufHttpContent(state, ctx, buf, msg.content(), msg.trailingHeaders(), out);
+            if (copyContent) {
+                assert state == ST_CONTENT_NON_CHUNK;
+                // verify if the estimation we made before was correct to hold the content,
+                // otherwise fallback to the existing path: subsequent attempts will likely be more accurate
+                if (buf.maxFastWritableBytes() >= content.readableBytes()) {
+                    buf.writeBytes(content);
+                    out.add(buf);
+                    return;
+                }
+            }
```

Comment:
Is this necessary? It seems like we'll get this behavior by following the main path. Specifically we'll take the route [205](https://github.com/netty/netty/blob/065a3478e9b1f8f0ebd09cf06cf394ff71fc5597/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java#L205) -> [334](https://github.com/netty/netty/blob/065a3478e9b1f8f0ebd09cf06cf394ff71fc5597/codec-http/src/main/java/io/netty/handler/codec/http/HttpObjectEncoder.java#L334). I think your more intelligent size estimation is all that is really necessary.

Reply:
Yeah, probably I could just go through this and let the original encode method to perform the same decision; I preferred to inline it to avoid messing up with the branch prediction on the encode method, polluting it and given that here we already know that is non chunked content, saving to recheck for it.

Reply:
Unless we have data to support a meaningful speedup in the context of encoding a HTTP message I suggest we skip it.